File: 0022-generate-parentheses/0022-generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis=function(n){

    let res=[];
    let tmp=[];

    function fun(open,close){

        if(open===n && close===n){
            // Jab Limit wla hit hojae
            res.push(tmp.join(""));
            return;
        }

        // Open wla Parenthesis k liye
        if(open<n){
            tmp.push('(');
            fun(open+1,close);
            tmp.pop();
        }

        // Close Wala Parenthesis lelo
        if(close<open){
            tmp.push(')');
            fun(open,close+1);
            tmp.pop();
        }
    }

    // Recursive Function Call kardo
    fun(0,0);

    return res;
};

File: 0022-generate-parentheses/0022-generate-parentheses.go
func generateParenthesis(n int) []string {

	var res []string
	tmp:= ""

	var fun func(int,int)

	fun = func(open,close int){

		if open==n && close==n{
			// Jab Limit wla hit hojae
			res=append(res,tmp)
			return
		}

		// Open wla Parenthesis k liye
		if open<n{
			tmp+="("
			fun(open+1,close)
			tmp=tmp[:len(tmp)-1]
		}

		// Close Wala Parenthesis lelo
		if close<open{
			tmp+=")"
			fun(open,close+1)
			tmp=tmp[:len(tmp)-1]
		}
	}

	// Recursive Function Call kardo
	fun(0,0)

	return res
}